Make all ra dec

rubyvanrooyen opened the pull request:
In order to expand the observation script to include tracking of solar body targets using AstroKAT, significant cleanup had to be done first to the library
These changes need to be incorporated on site first before the additional observation type and need to be merged as a priority before all new devel.

Please note, targets module will require `astropy` to be installed as a systematic move away from `ephem` and python 2.

Updated items are:
* Major cleanup of fundamental observation script to ensure consistent definition of targets and observations
* Removal of target construction from main observation script into stand alone `targets` module that also handles coordinate conversion
* Removal of katpoint_target construction from `utility` script to `targets` module
* Allow YAML input to handle galactic and scans using horizontal coordinates with tracking by converting these targets to equatorial (ra, dec) when observation starts.
* Update scan target definition to bring it in line with normal tracking target specification. All targets will now require both coordinate and duration specification, no more special case scan duration implementation with slightly different meanings.
*  Debug `scans` module to work with the new updated noise diode implementations
* Revamp all the tests to bring them in line with the updated implementations and to verify that the fundamental observation functionality works correctly using the new `targets` module.
* Add new tests to ensure `scans` module does not silently break again
* Update example YAML template to reflect updated implementation
* Update WIKI pages to ensure user docs stay in line with newer implementations

ajoubertza commented on astrokat/test/check_nd_units.sh:
```diff
@@ -0,0 +1,7 @@
+python -m unittest astrokat.test.test_nd_timings.TestAstrokatYAML.test_nd_pattern_ants
+python -m unittest astrokat.test.test_nd_timings.TestAstrokatYAML.test_nd_pattern_plus_off
+python -m unittest astrokat.test.test_nd_timings.TestAstrokatYAML.test_nd_trigger_long
+python -m unittest astrokat.test.test_nd_timings.TestAstrokatYAML.test_nd_trigger_short
```
These scripts may be useful, but are a bit of extra maintenance.  E.g., when we add more tests.

rubyvanrooyen replied:
No, wrong branch, these bash scripts are for me -- I have a devel branch that I stick this type of stuff in
Sorry this was sloppy, I cleaned up the tests

ajoubertza commented on astrokat/test/test_offline_observe.py:
```diff
@@ -30,8 +30,8 @@ def test_targets_sim(self):
-        self.assertIn("target2_gal observed for 10.0 sec", result)
+        self.assertIn("target1_gal observed for 10.0 sec", result)
+        self.assertIn("target2_azel observed for 10.0 sec", result)
 
```
Shouldn't we include some verification for target3?
```
name=target3_azel, azel=10 50, tags=target, duration=10.0, type=scan
```

rubyvanrooyen replied:
Indeed, added a log message and validation test for scan target
```
self.assertIn("target3_azel, tags=radec target, 15:56:30.60 8:51:02.5", result)
```

ajoubertza reviewed (commented):
Nice improvements.  I've only managed a partial review, but here are some comments so far.  More to follow.

rubyvanrooyen commented on astrokat/observatory.py:
```diff
@@ -6,13 +6,13 @@
 import json
 import ephem
 import numpy
+import astrokat.targets
```
a yes, this is the spagetti code I forgot to fix -- corrected

rubyvanrooyen commented on astrokat/observe_main.py:
```diff
@@ -910,6 +846,19 @@ def main(args):
+                start_ts = datetime2timestamp(obs_time_info["start_time"])
+        obs_targets = targets.read(obs_loop["target_list"],
+                                   timestamp=start_ts)
+        opts.obs_plan_params['observation_loop'][cntr]['target_list'] = obs_targets
```
This update makes it nice and clean, thank you

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+    return name, target
+
+
+def build_target(target_dict, timestamp=None):
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+# celestial targets, horizontal targets, galactic targets, solar system bodies
+# where the solar system bodies are planets and moons in our solar system that
+# do not follow standard celestial orbits.
+coords = ["radec", "azel", "gal"]
```
:+1:

I prefer to use `SUPPORTED_COORDINATE_TYPES`, it is a little more explicit

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+
+# -- library function --
+def tuple2str_(ra_float, dec_float):
```
Function name cleanup: `tuple2str_` -> `radec_to_string`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+    return ra_str, dec_str
+
+
+def get_radec_(pointing,
```
Function name cleanup: `get_radec_` -> `radec_from_pointing_object`

ajoubertza commented on astrokat/test/test_obs/targets-sim.yaml:
```diff
@@ -1,10 +1,11 @@
   product: c856M4k
 durations:
-  start_time: 2018-07-23 18:00:00Z
+  start_time: 2018-07-23 18:00:00
```
Fixed in #103

rubyvanrooyen replied:
The reason for removing the explicit use of Z is that it snuck in as an old C&P of start times of schedule blocks during debugging and development.
Astronomers think in UTC, so the Z is irrelevant, the start times should always be UTC.
I thus simply decided to remove the explicit Z to standardise the tests input yamls, and I missed  a few of the older files that I forgot to update.

I updated and implemented the correction as suggested by @ajoubertza, it makes much more sense to start consistent with `katpoint.Timestamp`, but on my tests cannot reproduce and error to check, so will have to rely on the CAM AQF system to see if the updates work.

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+    return location
+
+
+def eq2hor(ra_hms, dec_dms, location, timestamp,
```
Function name cleanup: `eq2hor` -> `radec_to_altaz`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+        return tgt_altaz.alt.deg, tgt_altaz.az.deg
+
+
+def hor2eq(az_deg, el_deg, location, timestamp,
```
Function name cleanup: `hor2eq` -> `altaz_to_radec`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                      as_string=as_string)
+
+
+def gal2eq(l_deg, b_deg,
```
Function name cleanup: `gal2eq` -> `galactic_to_radec`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                      as_string=as_string)
+
+
+def solar2eq(body, location, timestamp,
```
Function name cleanup: `solar2eq` -> `solarbody_to_radec`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+
+# -- coordinate conversion utilities --
+def mkat_locale(observer):
```
Function name cleanup: `mkat_locale` -> `observer_as_earth_location`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+    Returns
+    -------
+    list: [alt, az] horizontal coordinates in degrees
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+    Returns
+    -------
+    list: [ra, dec] equatorial coordinates in degrees
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+    Returns
+    -------
+    list: [ra, dec] equatorial coordinates in degrees
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+    Returns
+    -------
+    list: [ra, dec] equatorial coordinates in degrees
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                      as_string=as_string)
+
+
+def get_radec_coords(target_str, timestamp=None, convert_azel=False):
```
I settled on `get_coordinates_as_radec` since this best describe the functionality that is implemented

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                                unit=(u.hourangle, u.deg),
+                                frame='icrs')
+        ra_hms, dec_dms = tuple2str_(pointing.ra, pointing.dec)
+        tgt_coord = '{} {}'.format(str(ra_hms), str(dec_dms))
```
The `str()` is to convert the returned numpy unicode string to python flat string, but that should be handled in the function and I corrected it.

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                                 b_deg,
+                                 as_string=True)
+        tgt_type = "radec"
+        tgt_coord = '{} {}'.format(str(ra_hms), str(dec_dms))
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                                 timestamp,
+                                 as_string=True)
+        tgt_type = "radec"
+        tgt_coord = '{} {}'.format(str(ra_hms), str(dec_dms))
```
:+1:

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+# -- coordinate conversion utilities --
+
+
+def unpack_target(target_str, timestamp=None):
```
I think the name that best describe the function will be `parse_target_string`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+    return target
+
+
+def katpoint_target(target_str=None,
```
That is true, it is technically the` katpoint.Target.description`, but functionally it builds the string that is taken as input by `katpoint.Target`, so I think `katpoint_target_string` will be most appropriate

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+    return name, target
+
+
+def build_target(target_dict, timestamp=None):
```
The `build_target` function literally constructs the a tuple with values as expected by the `tgt_desc` `recarray` definition.
It does not actually populate a record, rather return a tuple of values in the expected format, which is appended to the `recarray`.
I feel more comfortable just extending the name to something like `build_target_tuple`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+
+    """
+    ntargets = len(target_items)
+    target_list = np.recarray(ntargets, dtype=tgt_desc)
```
I can understand that the use of the word `list` may  cause some confusion, since it is not a list as in the techinical sense, but rather and array of records.
So I am going to suggest again simply being explicit and updating it to: `target_list` -> `target_recarray`

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+        # build astrokat target info from dict definition
+        target_dict = unpack_target(target_item, timestamp=timestamp)
+        # pack target dictionary in observation ready target rec-array
+        target_ = build_target(target_dict,
```
Updated unfinished name `target_` to explicit `target_tuple` describing the expected return type

rubyvanrooyen commented on astrokat/targets.py:
```diff
@@ -0,0 +1,436 @@
+                      format(target_list.dtype))
+    return target_list
+
+
```
@ajoubertza thank you for all the naming suggestions. I agree, the names in the scripts needed a lot of attention.
I went through and updated, plus added comments, please have a look

ajoubertza commented on astrokat/simulate.py:
```diff
@@ -154,9 +154,9 @@ def __init__(self, kat, **kwargs):
-                self.start_time = datetime2timestamp(
-                    self.obs_params["durations"]["start_time"]
-                )
+                self.start_time = katpoint.Timestamp(
+                    str(self.obs_params["durations"]["start_time"])
+                ).secs
```
Sorry, I know I mentioned changing this in a previous comment, but I suggest we don't change it.  When I was working on #103, I saw that `datetime2timestamp` and `timestamp2datetime` as used consistently across astrokat, so better to leave it that way.  In #103 I added tests that ensure `katpoint.Timestamp` conversion agrees with the astrokat functions instead.

rubyvanrooyen replied:
Not a problem at all, I rolled back the `katpoint.Timestamp` updates to the original `datetime2timestamp` implementation

ajoubertza reviewed (changes requested):
Thanks for all the improvements.  One item to change (sorry).  And Python 3 tests need to be fixed.

rubyvanrooyen commented:
@ajoubertza and @rohanschwartz I have cleaned up the naming conventions and cleaned up the test scripts, also moved the unnecessary test scripts away.
Please have a look

ajoubertza reviewed (approved):
Looks good.  Thanks @rubyvanrooyen.
